src/Agent.Plugins/PipelineCache/PipelineCacheServer.cs

```diff
@@ -42,32 +52,150 @@ public class PipelineCacheServer
+                        {
+                            process.Kill();
+                            ExceptionDispatchInfo.Capture(e).Throw();
+                        }
 
-                // Send results to CustomerIntelligence
-                context.PublishTelemetry(area: PipelineArtifactConstants.AzurePipelinesAgent, feature: PipelineArtifactConstants.PipelineCache, record: uploadRecord);
-                context.PublishTelemetry(area: PipelineArtifactConstants.AzurePipelinesAgent, feature: PipelineArtifactConstants.PipelineCache, record: cacheRecord);
+                        var output = new List<string>();
+                        Func<string, StreamReader, Task> readLines = (prefix, reader) => Task.Run(async () =>
+                        {
+                            string line;
+                            while (null != (line = await reader.ReadLineAsync()))
+                            {
+                                lock (output)
+                                {
+                                    output.Add($"{prefix}{line}");
+                                }
+                            }
+                        });
+                        Task readStdOut = readLines("stdout: ", process.StandardOutput);
+                        Task readStdError = readLines("stderr: ", process.StandardError);
+
+                        // Our goal is to always have the process ended or killed by the time we exit the function.
+                        try
+                        {
+                            using (cancellationToken.Register(() => process.Kill()))
+                            {
+                                // readStdOut and readStdError should only fail if the process dies
+                                // processTcs.Task cannot fail as we only call SetResult on processTcs
+                                await Task.WhenAll(readStdOut, readStdError, processTcs.Task);
+                            }
+
+                            int exitCode = await processTcs.Task;
+
+                            if (exitCode == 0)
+                            {
+                                context.Output($"Process exit code: {exitCode}");
+                                foreach (string line in output)
+                                {
+                                    context.Output(line);
+                                }
+                            }
+                            else
+                            {
+                                throw new Exception($"Process returned non-zero exit code: {exitCode}");
+                            }
+                        }
+                        catch (Exception e)
+                        {
+                            // Delete archive file.
+                            if(File.Exists(archiveFile))
+                            {
+                                File.Delete(archiveFile);
+                            }
```

Comment:
needs to be encased into try/catch where catch does nothing:

try
{
File.Delete(archiveFile);
}
catch { }